File: app/src/main/java/com/kinwatt/powermeter/ui/fragments/DeviceFragment.java
package com.kinwatt.powermeter.ui.fragments;

import android.content.Context;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.kinwatt.powermeter.R;
import com.kinwatt.powermeter.data.SensorData;
import com.kinwatt.powermeter.data.provider.SensorProvider;

import java.util.ArrayList;
import java.util.List;

public class DeviceFragment extends Fragment {

    private OnListFragmentInteractionListener listener;

    private DeviceRecyclerViewAdapter adapter;

    private List<SensorData> items = new ArrayList<>();
    private SensorProvider provider;

    public DeviceFragment() { }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_device_list, container, false);

        provider = SensorProvider.getProvider(getActivity());
        items = provider.getAll();

        RecyclerView recyclerView = (RecyclerView) view;
        recyclerView.setLayoutManager(new LinearLayoutManager(view.getContext()));
        recyclerView.setAdapter(adapter = new DeviceRecyclerViewAdapter(items, listener));

        return view;
    }

    public void add(SensorData sensorData) {
        provider.add(sensorData);
        adapter.notifyItemInserted(items.size() - 1);
    }

    public boolean contains(SensorData sensorData) {
        return items.contains(sensorData);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnListFragmentInteractionListener) {
            listener = (OnListFragmentInteractionListener) context;
        } else {
            //throw new RuntimeException(context.toString() + " must implement OnListFragmentInteractionListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        listener = null;
    }

    public interface OnListFragmentInteractionListener {
        void onListFragmentInteraction(SensorData sensorData);
    }
}

File: app/src/main/java/com/kinwatt/powermeter/ui/fragments/DeviceFragment.kt
package com.kinwatt.powermeter.ui.fragments

import android.content.Context
import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.kinwatt.powermeter.R
import com.kinwatt.powermeter.data.SensorData
import com.kinwatt.powermeter.data.provider.SensorProvider
import com.kinwatt.powermeter.databinding.FragmentDeviceBinding

import java.util.ArrayList

class DeviceFragment : Fragment() {

    private var listener: OnListFragmentInteractionListener? = null

    private val deviceAdapter: DeviceRecyclerViewAdapter by lazy { DeviceRecyclerViewAdapter(items, listener) }

    private var items: List<SensorData> = ArrayList()
    private lateinit var provider: SensorProvider
    private lateinit var binding: FragmentDeviceBinding

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = FragmentDeviceBinding.inflate(inflater, container, false)
        val view = binding.root

        provider = SensorProvider.getProvider(activity!!)
        items = provider.all

        val recyclerView = view as RecyclerView
        recyclerView.layoutManager =
            LinearLayoutManager(view.getContext())
        recyclerView.adapter = deviceAdapter

        return view
    }

    fun add(sensorData: SensorData) {
        provider.add(sensorData)
        deviceAdapter.notifyItemInserted(items.size - 1)
    }

    operator fun contains(sensorData: SensorData): Boolean {
        return items.contains(sensorData)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is OnListFragmentInteractionListener) {
            listener = context
        } else {
            //throw new RuntimeException(context.toString() + " must implement OnListFragmentInteractionListener");
        }
    }

    override fun onDetach() {
        super.onDetach()
        listener = null
    }

    interface OnListFragmentInteractionListener {
        fun onListFragmentInteraction(sensorData: SensorData)
    }
}
